File: Interviewbit/linked-lists/remove-nth-node-from-list-end.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
ListNode* Solution::removeNthFromEnd(ListNode* A, int B) {
    ListNode *p=A,*q=A;
    while(B--){
        q=q->next;
        if(q==NULL) return A->next;
    }
    ListNode* r=A;
    while(q!=NULL){
        r=p;
        p=p->next;
        q=q->next;
    }
    r->next=p->next;
    delete(p);
    return A;
}

File: Interviewbit/linked-lists/remove-nth-node-from-list-end.java
/**
 * Definition for singly-linked list.
 * class ListNode {
 *     public int val;
 *     public ListNode next;
 *     ListNode(int x) { val = x; next = null; }
 * }
 */
public class Solution {
    public ListNode removeNthFromEnd(ListNode A, int B) {
        ListNode p=A;
        int count=0;
        while(B>count){
            count++;
            p=p.next;
            if(p==null) return A.next;
        }
        ListNode prev=A,q=A;
        while(p!=null){
            prev=q;
            q=q.next;
            p=p.next;
        }
        prev.next=q.next; //prev.next=q.next;
        return A;
    }
}
